### PR TITLE
add dns support in`ipfs p2p forward` and refactor code

### LINKED ISSUE #5524: `ipfs p2p forward` should accept full multiaddresses as target (not only /ipfs/<peerID>)
#### Version information: 

master

#### Type:

Feature

#### Description:

`ipfs p2p forward` takes an `*/ipfs/<peerID>` multi-address as target. ~~This assumes the given peer is routable.~~

However, sometimes it would be useful to do `ipfs p2p forward <proto> /dnsaddr/mypeer`  ~~or `ipfs p2p forward <proto> /ip4/1.2.4.5/tcp/23131/ipfs/<peerID>`~~. In these case, the command should ~~automatically add the peer addresses to the peerstore so that they can be connected (right now, a manual `swarm connect` is required in advance).~~ resolve the addresses.

~~This simplifies connecting to libp2p peers which are not part of the ipfs network or, particularly, the ipfs dht.~~

### PR DESCRIPTION
**Pull request by kjzz**
Fix #5524 

License: MIT
Signed-off-by: Kejie Zhang <601172892@qq.com>

### CODE REVIEW

**Comment by kjzz** [comment]
Hey @Stebalien , i have finished about this pr.If you have any time,please help me review it.Thx a lot.

File: core/commands/p2p.go [hunk 1]
```diff
@@ -19,6 +19,8 @@ import (
 	ma "gx/ipfs/QmYmsdtJ3HsodkePE3eU3TsCaP2YvPZJ4LoXnNkDE5Tpt7/go-multiaddr"
 	"gx/ipfs/QmZNkThpqfVXs9GNbexPrfBbXSLNYeKrE7jwFM2oqHbyqN/go-libp2p-protocol"
 	pstore "gx/ipfs/QmfAQMFpgDU2U4BXG64qVr8HSiictfWvkSBz7Y2oDj65st/go-libp2p-peerstore"
+	madns "gx/ipfs/QmfXU2MhWoegxHoeMd3A2ytL2P6CY4FfqGWc23LTNWBwZt/go-multiaddr-dns"
+	"time"
```
Comment (magik6k) [inline core/commands/p2p.go]: Move this to the top

File: test/sharness/t0180-p2p.sh [hunk 2]
```diff
@@ -73,10 +73,18 @@ test_server_to_client() {
...
 spawn_sending_server
 
-test_expect_success 'S->C Setup client side' '
+test_expect_success 'S->C(/ipfs/peerID) Setup client side' '
   ipfsi 1 p2p forward /x/p2p-test /ip4/127.0.0.1/tcp/10102 /ipfs/${PEERID_0} 2>&1 > dialer-stdouterr.log
 '
 
+test_expect_success 'S->C Setup(dnsaddr/addr) client side' '
+  ipfsi 1 p2p forward /x/p2p-test /ip4/127.0.0.1/tcp/10103 /dnsaddr/bootstrap.libp2p.io/ 2>&1 > dialer-stdouterr.log
+'
```
Comment (magik6k) [inline test/sharness/t0180-p2p.sh]: It would be great if we found a nicer way to test this

File: test/sharness/t0180-p2p.sh [hunk 3]
```diff
@@ -73,10 +73,18 @@ test_server_to_client() {
...
+
+test_expect_success 'S->C Setup(dnsaddr/addr/ipfs/peerID) client side' '
+  ipfsi 1 p2p forward /x/p2p-test /ip4/127.0.0.1/tcp/10104 /dnsaddr/bootstrap.libp2p.io/ipfs/${PEERID_0}  2>&1 > dialer-stdouterr.log
```
Comment (magik6k) [inline test/sharness/t0180-p2p.sh]: Should add a check to `check_test_ports` for 10104 above

File: core/commands/p2p.go [hunk 4]
```diff
@@ -129,14 +141,45 @@ Example:
...
 	},
 }
 
+// parseIpfsAddr is a function that takes in addr string and return ipfsAddrs
+func parseIpfsAddr(addr string) ([]ipfsaddr.IPFSAddr, error) {
+	mutiladdr, err := ma.NewMultiaddr(addr)
+	if err != nil {
+		return nil, err
+	}
+	if strings.Contains(mutiladdr.String(), "/ipfs/Qm") {
```
Comment (magik6k) [inline core/commands/p2p.go]: I wouldn't check for the `Qm` prefix as that won't work with Ed keys. Also, we want to migrate form `/ipfs/` to `/p2p/` in near future so for compatibility this should just be `if _, err := multiaddr.ValueForProtocol(ma.P_IPFS); err == nil {`

File: core/commands/p2p.go [hunk 5]
```diff
@@ -49,6 +51,16 @@ type P2PStreamsOutput struct {
...
 
+const (
+	allowCustomProtocolOptionName = "allow-custom-protocol"
+	allOptionName                 = "all"
+	protocolOptionName            = "protocol"
+	listenAddressOptionName       = "listen-address"
+	targetAddressOptionName       = "target-address"
+)
+
+var connectionTimeout = 10 * time.Second
```
Comment (magik6k) [inline core/commands/p2p.go]: This doesn't appear to be used anywhere but `parseIpfsAddr`, so maybe call it `resolveTimeout`?

File: test/sharness/t0180-p2p.sh (hunk 2 above)
Comment (kjzz) [reply test/sharness/t0180-p2p.sh]: I just imitated the test `ipfsi 1 p2p forward /x/p2p-test /ip4/127.0.0.1/tcp/10102 /ipfs/${PEERID_0} 2>&1 > dialer-stdouterr.log` . And use `test_expect_success`  to judge the test success.

**Comment by kjzz** [comment]
Thx a lot for your great advice @magik6k . I have update my pr please help me review again.

File: test/sharness/t0180-p2p.sh (hunk 2 above)
Comment (kjzz) [reply test/sharness/t0180-p2p.sh]: And i had add `test_must_be_empty dialer-stdouterr.log` to judge it succeed

Comment (magik6k) [reply test/sharness/t0180-p2p.sh]: You can also call `ipfsi 1 p2p ls` and check if it's output contains resolved listener (`grep '/x/p2p-test /ip4/127.0.0.1/tcp/10104 /'` or something like that)

Comment (kjzz) [reply test/sharness/t0180-p2p.sh]: @magik6k Thx for advice , i have update the test. Please help me review it again,

**Pull request merged**